python/activator/logger.py

```diff
@@ -0,0 +1,57 @@
+# This file is part of prompt_prototype.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["GCloudStructuredLogFormatter"]
+
+import json
+import logging
+
+
+class GCloudStructuredLogFormatter(logging.Formatter):
+    """A formatter that can be parsed by the Google Cloud logging agent.
+
+    The formatter's output is a JSON-encoded message containing keywords
+    recognized by the logging agent.
+
+    Parameters
+    ----------
+    labels : `dict` [`str`, `str`]
+        Any metadata that should be attached to the log. See ``LogEntry.labels``
+        in Google Cloud REST API documentation.
+    """
+    def __init__(self, labels=None):
+        super().__init__()
+
+        if labels:
+            self._labels = labels
+        else:
+            self._labels = {}
+
+    def format(self, record):
+        # Call for side effects only; ignore result.
+        super().format(record)
+
+        entry = {
+            "severity": record.levelname,
+            "logging.googleapis.com/labels": self._labels,
+            "message": record.getMessage(),
+        }
```

Comment:
Ironically, I had a commit like that, which I dropped as unnecessary complexity. :confused: 

No, I don't think we need the labels for anything else; my reasoning was that any `"labels": {}` in the output would just get silently dropped by Google, so there was no need to clean them out.

Reply:
Ah, alright, that makes sense. I had just been thinking that, if this was run often and the labels were often empty, it would slow things down when it didn't need to. Would we expect the number of times this gets called to make create empty labels and dropping them have a noticeable effect?

Reply:
Probably not? The activator code adds an instrument label now, and I'd expect any future evolution to be towards more labels, not less.